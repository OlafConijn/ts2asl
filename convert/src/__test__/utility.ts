import { writeFileSync } from "fs";
import { createCompilerHostFromFile } from "../compiler-host";
import { ConvertedStateMachine, ConvertedStateMachineWithDiagnostics, Converter } from "../convert";

export const runConvertForTest = (filename: string, stateMachineName: string = "main"): ConvertedStateMachine => {

  const host = createCompilerHostFromFile(
    `src/__test__/resources/${filename}.ts`
  );
  const converter = new Converter(host);
  const converted = converter
    .convert(true)
    .stateMachines.map(x => x as ConvertedStateMachineWithDiagnostics)
    .find(x => x.name === stateMachineName);

  if (!converted) throw Error("did not find state machine called " + stateMachineName);

  writeFileSync(
    `src/__test__/resources/output/${filename}-asllib.ts`,
    converted.transformedCode ?? ""
  );
  writeFileSync(
    `src/__test__/resources/output/${filename}-i-asl.json`,
    JSON.stringify(converted.iasl, function (this: any, key: string, val: any) { return key === "parentScope" ? undefined : val }, 2) ?? ""
  );
  writeFileSync(
    `src/__test__/resources/output/${filename}-asl.json`,
    JSON.stringify(converted.asl, null, 2) ?? ""
  );
  return converted;
}