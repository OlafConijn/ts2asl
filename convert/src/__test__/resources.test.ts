import { readdirSync, readFile, statSync, } from "fs";
import path from "path";
import { runConvertForTest } from "./utility";

test("when converting test-resources", () => {
  const dir = "src/__test__/resources";
  readdirSync(dir).forEach(file => {
    if (statSync(dir + "/" + file).isDirectory()) return;
    if (!file.endsWith(".ts")) return;
    if (file !== "account-creation.ts") return;

    const converted = runConvertForTest(file.substring(0, file.length - 3));
    expect((converted as any).transformedCode).toMatchSnapshot();
    expect((converted as any).iasl).toMatchSnapshot();
    expect(converted.asl).toMatchSnapshot();
  });
});