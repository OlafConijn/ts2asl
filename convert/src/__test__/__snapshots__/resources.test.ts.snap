// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when converting test-resources 1`] = `
"import * as asl from \\"@ts2asl/asl-lib\\";

export const replayPrefixer = asl.deploy.asLambda((input: StateMachineInput) => { return [\\"\\"] });

//option 1: native typescript
// export const main = asl.deploy.asStateMachine(async (input: StateMachineInput) => {
//   const result = await replayPrefixer(input);
//   for (const prefix of result) {
//     await asl.nativeSfnStartExecution({
//       parameters: {
//         input: asl.states.format(\\"{}\\", prefix),
//         stateMachineArn: asl.deploy.getParameter(\\"stateMachineArn\\"),
//       },
//     });
//   }
// });

//option 2: asl lib typescript
export const main = asl.deploy.asStateMachine(async (input: StateMachineInput) =>{
    const result = await asl.typescriptInvoke({
        name: \\"replayPrefixer(input)\\",
        resource: replayPrefixer,
        parameters: () => input,
        comment: \\"replayPrefixer(input)\\"
    });
    await asl.map({
        maxConcurrency: 5,
        items: result,
        iterator: (prefix) => asl.typescriptInvoke({
            name: \\"replayWorker({ prefix })\\",
            resource: replayWorker,
            parameters: () => ({ prefix }),
            comment: \\"replayWorker({ prefix })\\"
        })
    });
});

interface StateMachineInput {
  startDate: string;
  endDate?: string;
}


export const replayWorker = asl.deploy.asStateMachine(async (input: ReplayWorkerInput) => {
  const objects = await asl.nativeS3ListObjectsV2({ parameters: { Prefix: input.prefix, Bucket: \\"preprod-metrics-bucket-us-east-1\\" } });
  const itemsWithKeys = objects.Contents!.filter((item) => item.Key);
  const keys = itemsWithKeys.map(x => x.Key);

  await asl.map({
    items: keys,
    maxConcurrency: 5,
    iterator: key => {
      const message = {
        Records: [
          {
            replay: true,
            eventSource: \\"aws:s3\\",
            awsRegion: \\"us-east-1\\",
            eventName: \\"ObjectCreated:Put\\",
            s3: {
              s3SchemaVersion: \\"1.0\\",
              bucket: {
                name: \\"preprod-metrics-bucket-us-east-1\\",
                arn: \\"arn:aws:s3:::preprod-metrics-bucket-us-east-1\\"
              },
              object: {
                key: key
              }
            }
          }
        ]
      };
      asl.nativeSNSPublish({
        parameters: {
          TopicArn: \\"arn:aws:sns:us-east-1:400780617693:CentralizedMetricsStream-preprod-CentralizedMetricsStreamTopic48052E47-1X7SNV6Y357H6\\",
          Subject: \\"Ingestion Replay S3\\",
          Message: asl.states.jsonToString(message)
        }
      })
    }
  });
});

interface ReplayWorkerInput {
  prefix: string;
}
"
`;

exports[`when converting test-resources 2`] = `
Object {
  "_syntaxKind": "statemachine",
  "contextArgumentName": undefined,
  "inputArgumentName": Object {
    "_syntaxKind": "identifier",
    "identifier": "input",
  },
  "statements": Array [
    Object {
      "_syntaxKind": "variable-assignment",
      "expression": Object {
        "_syntaxKind": "asl-task-state",
        "catch": undefined,
        "parameters": Object {
          "_syntaxKind": "identifier",
          "identifier": "input",
          "type": "object",
        },
        "resource": "lambda:replayPrefixer",
        "retry": Array [
          Object {
            "BackoffRate": 2,
            "ErrorEquals": Array [
              "Lambda.ServiceException",
              "Lambda.AWSLambdaException",
              "Lambda.SdkClientException",
            ],
            "IntervalSeconds": 2,
            "MaxAttempts": 6,
          },
        ],
        "source": "replayPrefixer(input)",
        "stateName": "replayPrefixer(input)",
      },
      "name": Object {
        "_syntaxKind": "identifier",
        "identifier": "result",
        "type": "object",
      },
      "stateName": "Assign result",
    },
    Object {
      "_syntaxKind": "asl-map-state",
      "items": Object {
        "_syntaxKind": "identifier",
        "identifier": "result",
        "type": "object",
      },
      "iterator": Object {
        "_syntaxKind": "block",
        "statements": Array [
          Object {
            "_syntaxKind": "variable-assignment",
            "expression": Object {
              "_syntaxKind": "literal-object",
              "properties": Object {
                "AWS_STEP_FUNCTIONS_STARTED_BY_EXECUTION_ID": Object {
                  "_syntaxKind": "identifier",
                  "identifier": "$$.Execution.Id",
                  "type": "unknown",
                },
                "prefix": Object {
                  "_syntaxKind": "identifier",
                  "identifier": "prefix",
                  "type": "string",
                },
              },
            },
            "name": Object {
              "_syntaxKind": "identifier",
              "compilerGenerated": true,
              "identifier": "sfn_input",
              "type": "unknown",
            },
          },
          Object {
            "_syntaxKind": "asl-task-state",
            "catch": undefined,
            "parameters": Object {
              "_syntaxKind": "literal-object",
              "properties": Object {
                "Input": Object {
                  "_syntaxKind": "asl-intrinsic-function",
                  "arguments": Array [
                    Object {
                      "_syntaxKind": "identifier",
                      "compilerGenerated": true,
                      "identifier": "sfn_input",
                      "type": "unknown",
                    },
                  ],
                  "function": "asl.states.jsonToString",
                },
                "StateMachineArn": Object {
                  "_syntaxKind": "literal",
                  "type": "string",
                  "value": "statemachine:replayWorker",
                },
              },
            },
            "resource": "arn:aws:states:::aws-sdk:sfn:startExecution",
            "retry": Array [
              Object {
                "BackoffRate": 2,
                "ErrorEquals": Array [
                  "Lambda.ServiceException",
                  "Lambda.AWSLambdaException",
                  "Lambda.SdkClientException",
                ],
                "IntervalSeconds": 2,
                "MaxAttempts": 6,
              },
            ],
            "source": "replayWorker({ prefix })",
            "stateName": "replayWorker({ prefix })",
          },
        ],
      },
      "maxConcurrency": 5,
    },
  ],
}
`;

exports[`when converting test-resources 3`] = `
Object {
  "StartAt": "Initialize",
  "States": Object {
    "Initialize": Object {
      "Next": "replayPrefixer(input)",
      "Parameters": Object {
        "vars.$": "$$.Execution.Input",
      },
      "ResultPath": "$",
      "Type": "Pass",
    },
    "Map": Object {
      "Comment": undefined,
      "End": true,
      "ItemsPath": "$.vars.result",
      "Iterator": Object {
        "StartAt": "Assign Sfn_input",
        "States": Object {
          "Assign Sfn_input": Object {
            "Comment": undefined,
            "Next": "replayWorker({ prefix })",
            "Parameters": Object {
              "AWS_STEP_FUNCTIONS_STARTED_BY_EXECUTION_ID.$": "$$.Execution.Id",
              "prefix.$": "$.vars.prefix",
            },
            "ResultPath": "$.tmp.sfn_input",
            "Type": "Pass",
          },
          "replayWorker({ prefix })": Object {
            "Catch": undefined,
            "Comment": "source: replayWorker({ prefix })",
            "End": true,
            "HeartbeatSeconds": undefined,
            "Parameters": Object {
              "Input.$": "States.JsonToString($.tmp.sfn_input)",
              "StateMachineArn": "statemachine:replayWorker",
            },
            "Resource": "arn:aws:states:::aws-sdk:sfn:startExecution",
            "Retry": Array [
              Object {
                "BackoffRate": 2,
                "ErrorEquals": Array [
                  "Lambda.ServiceException",
                  "Lambda.AWSLambdaException",
                  "Lambda.SdkClientException",
                ],
                "IntervalSeconds": 2,
                "MaxAttempts": 6,
              },
            ],
            "TimeoutSeconds": undefined,
            "Type": "Task",
          },
        },
      },
      "MaxConcurrency": 5,
      "ResultPath": "$.tmp.lastResult",
      "Type": "Map",
    },
    "replayPrefixer(input)": Object {
      "Catch": undefined,
      "Comment": "source: replayPrefixer(input)",
      "HeartbeatSeconds": undefined,
      "InputPath": "$.vars",
      "Next": "Map",
      "Resource": "lambda:replayPrefixer",
      "ResultPath": "$.vars.result",
      "Retry": Array [
        Object {
          "BackoffRate": 2,
          "ErrorEquals": Array [
            "Lambda.ServiceException",
            "Lambda.AWSLambdaException",
            "Lambda.SdkClientException",
          ],
          "IntervalSeconds": 2,
          "MaxAttempts": 6,
        },
      ],
      "TimeoutSeconds": undefined,
      "Type": "Task",
    },
  },
}
`;
